src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp

```diff
@@ -296,33 +360,27 @@ void dfcc_spec_functionst::to_spec_assigns_function(
   INVARIANT(
     function_pointer_contracts.size() == 0,
     "discovered function pointer contracts unexpectedly");
+
   utils.set_hide(function_id, true);
 }
 
-void dfcc_spec_functionst::to_spec_frees_function(
-  const irep_idt &function_id,
+void dfcc_spec_functionst::to_spec_frees_instructions(
+  const exprt &write_set_to_fill,
+  goto_programt &program,
   std::size_t &nof_targets)
 {
   // counts the number of calls to the `freeable` builtin
   std::size_t next_idx = 0;
-  auto &goto_function = goto_model.goto_functions.function_map.at(function_id);
-
-  // add __dfcc_set parameter
-  auto &set_symbol = utils.add_parameter(
-    function_id,
-    "__write_set_to_fill",
-    library.dfcc_type[dfcc_typet::WRITE_SET_PTR]);
-
-  Forall_goto_program_instructions(ins_it, goto_function.body)
+  Forall_goto_program_instructions(ins_it, program)
   {
     if(ins_it->is_function_call())
     {
       if(ins_it->call_function().id() != ID_symbol)
       {
         throw invalid_source_file_exceptiont(
           "Function pointer call '" +
-            from_expr(ns, function_id, ins_it->call_function()) +
-            "' in function '" + id2string(function_id) + "' is not supported",
+            from_expr(ns, "", ins_it->call_function()) +
+            "' are not supported in frees clauses",
```

Comment:
As above: consider using `format_to_string` instead, and the grammar seems broken.